src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -202,9 +284,173 @@ public Task LoadAsync()
 
         public Task UnloadAsync()
         {
-            _enabled = false;
+            lock (SyncObject)
```

Comment:
What is this protecting against?

Reply:
set _enabled, and finish _whenNominatedTask need be inside the same syncObject lock. Otherwise, WhenNominated may create the task completion source when _enabled is true, which will not be completed when UnloadAsync is called at the same time.

Logically, if LoadAsync is called after UnloadAsync, ignored data in the dataflow will not be processed again. But that is an existing problem. not sure whether we care in the real product.